providers/implementations/keymgmt/mlx_kmgmt.c

```diff
@@ -0,0 +1,764 @@
+                return 0;
+        }
+    }
+    if (pub == NULL && prv == NULL)
+        return 1;
+
+    sub_arg.pubenc = pub ? pub->data : NULL;
+    sub_arg.prvenc = prv ? prv->data : NULL;
+    sub_arg.pubcount = 0;
+    sub_arg.prvcount = 0;
+
+    selection = prv == NULL ? 0 : OSSL_KEYMGMT_SELECT_PRIVATE_KEY;
+    selection |= pub == NULL ? 0 : OSSL_KEYMGMT_SELECT_PUBLIC_KEY;
+    if (key->xinfo->group_name)
+        selection |= OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS;
+
+    /* Extract sub-component key material */
+    if (!export_sub(&sub_arg, selection, key))
+        return 0;
+
+    if ((pub != NULL && sub_arg.pubcount != 2)
+        || (prv != NULL && sub_arg.prvcount != 2))
+        return 0;
+
+    return 1;
+}
+
+static const OSSL_PARAM *mlx_kem_settable_params(void *provctx)
+{
+    static const OSSL_PARAM arr[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return arr;
+}
+
+static int mlx_kem_set_params(void *vkey, const OSSL_PARAM params[])
+{
+    MLX_KEY *key = vkey;
+    const OSSL_PARAM *p;
+    const void *pubenc = NULL;
+    size_t publen = 0;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
+        return 1;
+
+    /* Key mutation is reportedly generally not allowed */
+    if (mlx_kem_have_pubkey(key)) {
+        ERR_raise_data(ERR_LIB_PROV,
+                       PROV_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE,
+                       "keys cannot be mutated");
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        OPENSSL_free(key->propq);
+        key->propq = NULL;
+        if (!OSSL_PARAM_get_utf8_string(p, &key->propq, 0))
+            return 0;
+    }
+
+    if (publen != key->minfo->pubkey_bytes + key->xinfo->pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+
+    return load_keys(key, pubenc, publen, NULL, 0);
+}
+
+static int mlx_kem_gen_set_params(void *vgctx, const OSSL_PARAM params[])
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    const OSSL_PARAM *p;
+
+    if (gctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        OPENSSL_free(gctx->propq);
+        if ((gctx->propq = OPENSSL_strdup(p->data)) == NULL)
+            return 0;
+    }
+    return 1;
+}
+
+static void *mlx_kem_gen_init(int v, OSSL_LIB_CTX *libctx, int selection,
+                              const OSSL_PARAM params[])
+{
+    static const int minimal =
```

Comment:
Don't get why you made this a static const ...

Reply:
It is static data that never changes, so a static const is natural.